sharder/deterministic.go

```diff
@@ -291,20 +294,101 @@ func (d *DeterministicSharder) loadPeerList() error {
 	} else {
 		d.peerLock.RUnlock()
 	}
+	// the random partitioning may have significant imbalances, so we try to
+	// balance them by moving partitions from the most heavily loaded nodes to
+	// the least heavily loaded nodes. But it's only possible if the number of
+	// partitions per peer is > 1.
+	if partitionsPerPeer > 1 {
+		d.balanceProportions()
+	}
 	return nil
 }
 
 func (d *DeterministicSharder) MyShard() Shard {
 	return d.myShard
 }
 
-// WhichShard calculates which shard we want by keeping a list of partitions. Each
-// partition has a different hash value and a map from partition to a given shard.
-// We take the traceID and calculate a hash for each partition, using the partition
-// hash as the seed for the trace hash. Whichever one has the highest value is the
-// partition we use, which determines the shard we use.
-// This is O(N) where N is the number of partitions, but because we use an efficient hash,
-// (as opposed to SHA1) it executes in 1 uSec for 50 partitions.
+// find the min and max nodes in terms of the amount of the space they're responsible for
+// and return the ratio of the max to the min
+func (d *DeterministicSharder) getMinMax() (int, int, float64) {
+	totals := make([]float64, len(d.peers))
+	for i := 1; i < len(d.hashes); i++ {
+		totals[d.hashes[i-1].shardIndex] += float64(d.hashes[i].uhash-d.hashes[i-1].uhash) / float64(math.MaxUint64)
+	}
+	minix, maxix := 0, 0
+	for i := range totals {
+		if totals[i] < totals[minix] {
+			minix = i
+		}
+		if totals[i] > totals[maxix] {
+			maxix = i
+		}
+	}
+	return minix, maxix, totals[maxix] / totals[minix]
+}
```

Comment:
We need to remove this content from the PR

Reply:
Done.